ui/pages/editor.tsx

```diff
@@ -1,110 +1,254 @@
+  }
 
-  const codeFetcher = async (url) => {
-    const data = await fetch(url + window.location.search);
-    return await data.text();
-  };
+  get isDirty(): boolean {
+    return this._isDirty;
+  }
 
-  const initialCodeRoute = props.fileName ? path.join('/code/' + props.fileName) : '';
-  const initialCode = props.code;
-  const initialName = props.fileName;
+  set isDirty(newDirtyState: boolean) {
+    const wasDirty = this.isDirty;
+    if(wasDirty !== newDirtyState) {
+      this._isDirty = newDirtyState;
+      this.onIsDirtyChange(this);
+    }
+  }
+
+  public saveModel() {
+    this.lastSavedVersionId = this.model.getAlternativeVersionId();
+    this.isDirty = false;
+    this.onModelSave();
+  }
+}
+
+export default function EditorPage(props) {
+  const applicationId = props.applicationId;
 
   const [isEditorReady, setIsEditorReady] = useState(false);
   const valueGetter = useRef<() => string>();
   const editorRef = useRef();
-  const [name, setName] = useState(props.fileName);
   const [fileTree, setFileTree] = useState<IFileTreeNode[]>([]);
-  const [codeRoute, setCodeRoute] = useState(initialCodeRoute)
+  const [monacoInstance, setMonacoInstance] = useState(null);
+  const [monacoInstanceLoaded, setMonacoInstanceLoaded] = useState(false);
+  const [editorModels, setEditorModels] = useState<Map<string, EditorModel>>(new Map<string, EditorModel>());
+  const [fileNodeMap, setFileNodeMap] = useState({});
+  const [newFileName, setNewFileName] = useState("");
+  const [selectedFileKey, setSelectedFileKey] = useState("");
 
   // https://github.com/vercel/swr/issues/284 - using initialData for SSR has open bug
-  const { data: code, error: codeError } = useSWR(name ? codeRoute : null, codeFetcher, { initialData: name === initialName ? initialCode : undefined }); 
+  // TODO: confirm this still works as expected
+  const { data: code, error: codeError, isValidating: codeValidating } = useSWR(selectedFileKey ? ['/static', selectedFileKey] : null, codeFetcher, { onSuccess: onCodeFetchSuccess})
+
+  async function codeFetcher(baseUrl, fileKey) {
+    const url = path.join(baseUrl, fileKey);
+    const data = await fetch(url + window.location.search);
+    const code = await data.text();
+    return code;
+  };
+
+  // when we have our code, fetch the proper model or create a new model to mount the code to
+  function onCodeFetchSuccess(data) {
+    let fileName = selectedFileKey;
+    if (!(fileName in editorModels)) {
+      const model = createEditorModel(fileName, data);
+      editorModels[fileName] = model;
+      setEditorModels({...editorModels});
+    }
+    editorRef.current.setModel(editorModels[fileName].model);
+  }
 
   useEffect(() => {
-      const transformedData = _transformFileTreeData(props.fileTree);
-      setFileTree(transformedData);
-      if(props.fileName) {
-        handleInitOpenFile(props.fileName);
-      }
+      const initializeMonaco = async () => {
+        const monacoInst = await monaco.init();
```

Comment:
So the editor you get out of the box is of this type: https://microsoft.github.io/monaco-editor/api/interfaces/monaco.editor.istandalonecodeeditor.html - basically its a single editor with an underlying model (Get/SetModel). As you can see from the API, you don't have the ability to create a model (a model is basically an independent shell for the data in an editor). Importing `monaco` and running init() gives you access to https://microsoft.github.io/monaco-editor/api/modules/monaco.editor.html (see createModel). Now we can create models for each file using this.

Reply:
Great. I imagined we'd have to go that direction eventually. Good stuff